project/MetalsPlugin.scala

```diff
@@ -1,85 +1,112 @@
 import sbt._
 import sbt.Keys._
 import java.io._
+import sbt.Def
 
 object MetalsPlugin extends AutoPlugin {
   override def trigger = allRequirements
   override def requires = sbt.plugins.JvmPlugin
-  val metalsCompilerConfig =
-    taskKey[String]("Configuration parameters for autocompletion.")
-  val metalsSetup =
-    taskKey[Unit](
-      "Generate build metadata for completions and indexing dependency sources"
+  object autoImport {
+
+    val metalsCompilerConfig =
+      taskKey[String](
+        "String containing build metadata in properties file format."
+      )
+    val metalsWriteCompilerConfig =
+      taskKey[Unit](
+        "Generate build metadata for completions and indexing dependency sources"
+      )
+
+    lazy val semanticdbSettings = List(
+      addCompilerPlugin(
+        "org.scalameta" % "semanticdb-scalac" % SemanticdbEnable.semanticdbVersion cross CrossVersion.full
+      ),
+      scalacOptions += "-Yrangepos"
+    )
+
+    def metalsConfig(c: Configuration) = Seq(
+      metalsCompilerConfig := {
+        val props = new java.util.Properties()
+        props.setProperty(
+          "sources",
+          sources.value.distinct.mkString(File.pathSeparator)
+        )
+        props.setProperty(
+          "unmanagedSourceDirectories",
+          unmanagedSourceDirectories.value.distinct
+            .mkString(File.pathSeparator)
+        )
+        props.setProperty(
+          "managedSourceDirectories",
+          managedSourceDirectories.value.distinct
+            .mkString(File.pathSeparator)
+        )
+        props.setProperty(
+          "scalacOptions",
+          scalacOptions.value.mkString(" ")
+        )
+        props.setProperty(
+          "classDirectory",
+          classDirectory.value.getAbsolutePath
+        )
+        props.setProperty(
+          "dependencyClasspath",
+          dependencyClasspath.value
+            .map(_.data.toString)
+            .mkString(File.pathSeparator)
+        )
+        props.setProperty(
+          "scalaVersion",
+          scalaVersion.value
+        )
+        val sourceJars = for {
+          configurationReport <- updateClassifiers.value.configurations
+          moduleReport <- configurationReport.modules
+          (artifact, file) <- moduleReport.artifacts
+          if artifact.classifier.exists(_ == "sources")
+        } yield file
+        props.setProperty(
+          "sourceJars",
+          sourceJars.mkString(File.pathSeparator)
+        )
+        val out = new ByteArrayOutputStream()
+        props.store(out, null)
+        out.toString()
+      },
+      metalsWriteCompilerConfig := {
+        val filename = s"${thisProject.value.id}-${c.name}.properties"
+        val basedir = baseDirectory.in(ThisBuild).value / ".metals" / "config"
```

Comment:
I suggest to change subdirectory name to avoid confusion with the server LSP config. I can imagine that some users after reading that they can turn on experimental features in _some configuration_ may miss that it's in their editor, stumble upon some files in `.metals/config/` and try editing them 😄 

I suggest `.metals/buildinfo/`, but I don't care about particular name, just not `config`.

Also how about changing to nested structure: `.metals/buildinfo/subproject/configuration.properties`? It nicely corresponds to the new sbt slash-notation for scoping.

Reply:
Good point, I also like the nested structure. `buildinfo` is clearer than `config`

Reply:
I like it too. It's also what mill does and I find it very intuitive

Reply:
Cool. I'm going to change it in #230 then.